drivers/modem/quectel-bg9x.c

```diff
@@ -0,0 +1,1155 @@
+ * is that this function will be called right after the modem is
+ * powered on to do the stuff necessary to talk to the modem.
+ */
+static int modem_setup(void)
+{
+	int ret = 0, counter;
+	int rssi_retry_count = 0, init_retry_count = 0;
+
+	/* Setup the pins to ensure that Modem is enabled. */
+	pin_init();
+
+restart:
+
+	counter = 0;
+
+	/* stop RSSI delay work */
+	k_delayed_work_cancel(&mdata.rssi_query_work);
+
+	/* Let the modem respond. */
+	LOG_INF("Waiting for modem to respond");
+	ret = modem_at(&mctx, &mdata);
+	if (ret < 0) {
+		LOG_ERR("MODEM WAIT LOOP ERROR: %d", ret);
+		goto error;
+	}
+
+	/* Run setup commands on the modem. */
+	ret = modem_cmd_handler_setup_cmds(&mctx.iface, &mctx.cmd_handler,
+					   setup_cmds, ARRAY_SIZE(setup_cmds),
+					   &mdata.sem_response, MDM_REGISTRATION_TIMEOUT);
+	if (ret < 0) {
+		goto error;
+	}
+
+restart_rssi:
+
+	/* query modem RSSI */
+	modem_rssi_query_work(NULL);
+	k_sleep(MDM_WAIT_FOR_RSSI_DELAY);
+
+	/* Keep trying to read RSSI until we get a valid value - Eventually, exit. */
+	while (counter++ < MDM_WAIT_FOR_RSSI_COUNT &&
+	      (mctx.data_rssi >= 0 || mctx.data_rssi <= -1000)) {
+		modem_rssi_query_work(NULL);
+		k_sleep(MDM_WAIT_FOR_RSSI_DELAY);
+	}
+
+	/* Is the RSSI invalid ? */
+	if (mctx.data_rssi >= 0 || mctx.data_rssi <= -1000) {
+		rssi_retry_count++;
+
+		if (rssi_retry_count >= MDM_NETWORK_RETRY_COUNT) {
+			LOG_ERR("Failed network init. Too many attempts!");
+			ret = -ENETUNREACH;
+			goto error;
+		}
+
+		/* Try again! */
+		LOG_ERR("Failed network init. Restarting process.");
+		counter = 0;
+		goto restart_rssi;
+	}
+
+	/* Network is ready - Start RSSI work in the background. */
+	LOG_INF("Network is ready.");
+	k_delayed_work_submit_to_queue(&modem_workq,
+				       &mdata.rssi_query_work,
+				       K_SECONDS(RSSI_TIMEOUT_SECS));
+
+	/* Once the network is ready, activate PDP context. */
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+			     NULL, 0U, "AT+QIACT=1", &mdata.sem_response,
+			     MDM_CMD_TIMEOUT);
+
+	/* Retry or Possibly Exit. */
+	if (ret < 0 && init_retry_count++ < MDM_INIT_RETRY_COUNT) {
+		goto restart;
+	}
+
+error:
+	return ret;
+}
+
+static const struct socket_op_vtable offload_socket_fd_op_vtable = {
+	.fd_vtable = {
+		.read	= offload_read,
+		.write	= offload_write,
+		.close	= offload_close,
+		.ioctl	= offload_ioctl,
+	},
+	.bind		= NULL,
+	.connect	= offload_connect,
+	.sendto		= offload_sendto,
+	.recvfrom	= offload_recvfrom,
+	.listen		= NULL,
+	.accept		= NULL,
+	.sendmsg	= offload_sendmsg,
+	.getsockopt	= NULL,
+	.setsockopt	= NULL,
+};
+
+static struct net_offload modem_net_offload = {
+	.get = net_offload_dummy_get,
+};
+
+/* Setup the Modem NET Interface. */
+static void modem_net_iface_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct modem_data *data	 = dev->data;
+
+	/* Direct socket offload used instead of net offload: */
+	iface->if_dev->offload = &modem_net_offload;
```

Comment:
and we should not need these if NET_OFFLOAD is not set.

Just wondering if there is some side effect which requires that NET_OFFLOAD is set. If this is the case, then it sounds like a bug (because we do not use NET_OFFLOAD for anything). Can you try to remove the NET_OFFLOAD and see what happens?